Query memory leaks when closing server

```
2026-03-24T14:14:34.886+01:00	xtdb-statefulset-0	INFO	xtdb.main	Stopping node...
2026-03-24T14:14:35.060+01:00	xtdb-statefulset-0	INFO	xtdb.healthz	Healthz server stopped.
2026-03-24T14:14:40.182+01:00	xtdb-statefulset-0	WARN	xtdb.pgwire	Could not drain connections in time, force closing
2026-03-24T14:14:40.184+01:00	xtdb-statefulset-0	ERROR	o.apache.arrow.memory.BaseAllocator	Memory was leaked by query. Memory leaked: (8704) Allocator(pg-conn-272) 0/8704/20736/4831838208 (res/actual/peak/limit)

2026-03-24T14:14:40.188+01:00	xtdb-statefulset-0	WARN	xtdb.util	could not close java.lang.IllegalStateException: Memory was leaked by query. Memory leaked: (8704) Allocator(pg-conn-272) 0/8704/20736/4831838208 (res/actual/peak/limit) 
at org.apache.arrow.memory.BaseAllocator.close(BaseAllocator.java:504) 
at xtdb.util$close.invokeStatic(util.clj:90) 
at xtdb.util$close.invoke(util.clj:73) 
at xtdb.pgwire.Connection.close(pgwire.clj:134) 
at xtdb.util$try_close.invokeStatic(util.clj:99) 
at xtdb.util$try_close.invoke(util.clj:95) 
at clojure.core$run_BANG_$fn__8926.invoke(core.clj:7907) 
at clojure.core.protocols$iterator_reduce_BANG_.invokeStatic(protocols.clj:42) 
at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:52) 
at clojure.core.protocols$fn__8268.invokeStatic(protocols.clj:74) 
at clojure.core.protocols$fn__8268.invoke(protocols.clj:74) 
at clojure.core.protocols$fn__8207$G__8202__8220.invoke(protocols.clj:13) 
at clojure.core$reduce.invokeStatic(core.clj:6965) 
at clojure.core$run_BANG_.invokeStatic(core.clj:7902) 
at clojure.core$run_BANG_.invoke(core.clj:7902) 
at xtdb.util$try_close.invokeStatic(util.clj:103) 
at xtdb.util$try_close.invoke(util.clj:95) 
at xtdb.pgwire.Server.close(pgwire.clj:104) 
at xtdb.util$close.invokeStatic(util.clj:90) 
at xtdb.util$close.invoke(util.clj:73) 
at xtdb.util$close$close_all__2469$fn__2472.invoke(util.clj:78) 
at clojure.core.protocols$iterator_reduce_BANG_.invokeStatic(protocols.clj:42) 
at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:52) 
at clojure.core.protocols$fn__8260.invokeStatic(protocols.clj:74) 
at clojure.core.protocols$fn__8260.invoke(protocols.clj:74) 
at clojure.core.protocols$fn__8207$G__8202__8220.invoke(protocols.clj:13) 
at clojure.core$reduce.invokeStatic(core.clj:6965) 
at clojure.core$reduce.invoke(core.clj:6947) 
at xtdb.util$close$close_all__2469.invoke(util.clj:76) 
at xtdb.util$close.invokeStatic(util.clj:91) 
at xtdb.util$close.invoke(util.clj:73) 
at xtdb.pgwire$eval29864$fn__29865.invoke(pgwire.clj:2003) 
at clojure.lang.MultiFn.invoke(MultiFn.java:234) 
at integrant.core$try_run_action.invokeStatic(core.cljc:369) 
at integrant.core$try_run_action.invoke(core.cljc:367) 
at integrant.core$run_loop.invokeStatic(core.cljc:377) 
at integrant.core$run_loop.invoke(core.cljc:373) 
at integrant.core$reverse_run_BANG_.invokeStatic(core.cljc:397) 
at integrant.core$reverse_run_BANG_.invoke(core.cljc:391) 
at integrant.core$halt_BANG_.invokeStatic(core.cljc:666) 
at integrant.core$halt_BANG_.invoke(core.cljc:660) 
at integrant.core$halt_BANG_.invokeStatic(core.cljc:663) 
at integrant.core$halt_BANG_.invoke(core.cljc:660) 
at xtdb.node.impl$open_node$fn__26493.invoke(impl.clj:381) 
at xtdb.node.impl.Node.close(impl.clj:277) 
at xtdb.util$close.invokeStatic(util.clj:90) 
at xtdb.util$close.invoke(util.clj:73) 
at xtdb.main$start_node.invokeStatic(main.clj:142) 
at xtdb.main$start_node.invoke(main.clj:139) 
at xtdb.main$_main.invokeStatic(main.clj:325) 
at xtdb.main$_main.doInvoke(main.clj:278) 
at clojure.lang.RestFn.applyTo(RestFn.java:140) 
at clojure.lang.Var.applyTo(Var.java:707) 
at clojure.core$apply.invokeStatic(core.clj:667) 
at clojure.main$main_opt.invokeStatic(main.clj:515) 
at clojure.main$main_opt.invoke(main.clj:511) 
at clojure.main$main.invokeStatic(main.clj:665) 
at clojure.main$main.doInvoke(main.clj:617) 
at clojure.lang.RestFn.applyTo(RestFn.java:140) 
at clojure.lang.Var.applyTo(Var.java:707) 
at clojure.main.main(main.java:40)

2026-03-24T14:14:40.292+01:00	xtdb-statefulset-0	ERROR	xtdb.pgwire	Uncaught exception processing message xtdb.error.Interrupted: null 
at xtdb.error$interrupted.invokeStatic(error.clj:60) 
at xtdb.error$interrupted.invoke(error.clj:51) 
at xtdb.error$eval1038$fn__1039.invoke(error.clj:172) 
at xtdb.error$eval961$fn__962$G__952__969.invoke(error.clj:14) 
at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:277) 
at xtdb.pgwire$send_ex.invoke(pgwire.clj:271) 
at xtdb.pgwire$eval29066$fn__29068$fn__29071.invoke(pgwire.clj:680) 
at xtdb.pgwire$eval29066$fn__29068.invoke(pgwire.clj:674) 
at clojure.lang.MultiFn.invoke(MultiFn.java:234) 
at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1718) 
at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1712) 
at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1765) 
at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1736) 
at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1811) 
at xtdb.pgwire$connect.invoke(pgwire.clj:1769) 
at xtdb.pgwire$accept_loop$fn__29811$fn__29812$fn__29813.invoke(pgwire.clj:1858) 
at clojure.lang.AFn.run(AFn.java:22) 
at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) 
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) 
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) 
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) 
at java.base/java.lang.VirtualThread.run(VirtualThread.java:329) Caused by: java.lang.InterruptedException: null 
at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:98) 
at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:70) 
at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source) 
at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:48) 
at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source) 
at xtdb.database.Database.awaitTxBlocking(Database.kt:74) 
at xtdb.node.impl$await_msg_result.invokeStatic(impl.clj:78) 
at xtdb.node.impl$await_msg_result.invoke(impl.clj:77) 
at xtdb.node.impl.Node.executeTx(impl.clj:178) 
at xtdb.pgwire$cmd_commit$fn__29004.invoke(pgwire.clj:577) 
at xtdb.pgwire$cmd_commit.invokeStatic(pgwire.clj:566) 
at xtdb.pgwire$cmd_commit.invoke(pgwire.clj:553) 
at xtdb.pgwire$eval29066$fn__29068$fn__29071.invoke(pgwire.clj:678) ... 15 common frames omitted

2026-03-24T14:14:40.312+01:00	xtdb-statefulset-0	ERROR	xtdb.pgwire	Uncaught exception processing message xtdb.error.Fault: Socket closed 
at xtdb.error$fault.invokeStatic(error.clj:70) 
at xtdb.error$fault.invoke(error.clj:62) 
at xtdb.error$eval1034$fn__1035.invoke(error.clj:184) 
at xtdb.error$eval961$fn__962$G__952__969.invoke(error.clj:14) 
at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:277) 
at xtdb.pgwire$send_ex.invoke(pgwire.clj:271) 
at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1734) 
at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1712) 
at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1765) 
at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1736) 
at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1811) 
at xtdb.pgwire$connect.invoke(pgwire.clj:1769) 
at xtdb.pgwire$accept_loop$fn__29811$fn__29812$fn__29813.invoke(pgwire.clj:1858) 
at clojure.lang.AFn.run(AFn.java:22) 
at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) 
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) 
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) 
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) 
at java.base/java.lang.VirtualThread.run(VirtualThread.java:329) Caused by: java.net.SocketException: Socket closed 
at java.base/sun.nio.ch.NioSocketImpl.ensureOpenAndConnected(NioSocketImpl.java:163) 
at java.base/sun.nio.ch.NioSocketImpl.beginWrite(NioSocketImpl.java:362) 
at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:407) 
at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:440) 
at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:819) 
at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1195) 
at java.base/java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:125) 
at java.base/java.io.BufferedOutputStream.implFlush(BufferedOutputStream.java:252) 
at java.base/java.io.BufferedOutputStream.flush(BufferedOutputStream.java:240) 
at java.base/java.io.DataOutputStream.flush(DataOutputStream.java:131) 
at xtdb.pgwire.io.SocketFrontend.send_client_msg_BANG_(io.clj:136) 
at xtdb.pgwire.io$cmd_write_msg.invokeStatic(io.clj:78) 
at xtdb.pgwire.io$cmd_write_msg.invoke(io.clj:72) 
at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:286) 
at xtdb.pgwire$send_ex.invoke(pgwire.clj:271) 
at xtdb.pgwire$eval29066$fn__29068$fn__29071.invoke(pgwire.clj:680) 
at xtdb.pgwire$eval29066$fn__29068.invoke(pgwire.clj:674) 
at clojure.lang.MultiFn.invoke(MultiFn.java:234) 
at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1718) ... 12 common frames omitted

2026-03-24T14:14:40.313+01:00	xtdb-statefulset-0	INFO	xtdb.pgwire	Server stopped.
2026-03-24T14:14:40.357+01:00	xtdb-statefulset-0	INFO	xtdb.api.FlightSql	Flight SQL server stopped
```

--- a/core/src/main/clojure/xtdb/pgwire.clj
+++ b/core/src/main/clojure/xtdb/pgwire.clj
@@ -127,14 +127,18 @@
 
   Closeable
   (close [_]
+    (log/debug "Closing connection..." {:cid cid})
     (util/close frontend)
+
+    (doseq [portal (vals (:portals @conn-state))]
+      (util/close (:cursor portal)))
 
     (let [{:keys [server-state]} server]
       (swap! server-state update :connections dissoc cid))
 
     (util/close allocator)
 
-    (log/debug "Connection ended" {:cid cid})))
+    (log/debug "Connection closed" {:cid cid})))
 
 ;; best the server/conn records are opaque when printed as they contain mutual references
 (defmethod print-method Server [wtr o] ((get-method print-method Object) wtr o))

